// xgplayer-helper-codec/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UTF8 = exports.CompatHls = exports.Compat = exports.hevc = exports.avc = exports.ADTS = exports.AAC = undefined;

var _aac = require('./aac');

var _aac2 = _interopRequireDefault(_aac);

var _adts = require('./aac/adts');

var _adts2 = _interopRequireDefault(_adts);

var _avc = require('./avc');

var avc = _interopRequireWildcard(_avc);

var _hevc = require('./hevc');

var hevc = _interopRequireWildcard(_hevc);

var _compat = require('./compat');

var _compat2 = _interopRequireDefault(_compat);

var _hls = require('./compat/hls');

var _hls2 = _interopRequireDefault(_hls);

var _utf = require('./utf8');

var _utf2 = _interopRequireDefault(_utf);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.AAC = _aac2.default;
exports.ADTS = _adts2.default;
exports.avc = avc;
exports.hevc = hevc;
exports.Compat = _compat2.default;
exports.CompatHls = _hls2.default;
exports.UTF8 = _utf2.default;
